applications/FluidDynamicsApplication/python_scripts/fluid_dynamics_analysis.py

```diff
@@ -0,0 +1,232 @@
+
+        ## Stepping and time settings
+        self.end_time = self.project_parameters["problem_data"]["end_time"].GetDouble()
+
+        ## Writing the full ProjectParameters file before solving
+        if self.is_printing_rank and self.echo_level > 1:
+            with open("ProjectParametersOutput.json", 'w') as parameter_output_file:
+                parameter_output_file.write(self.project_parameters.PrettyPrintJsonString())
+
+
+    def _SetUpGiDOutput(self):
+        '''Initialize self.output as a GiD output instance.'''
+        self.have_output = self.project_parameters.Has("output_configuration")
+        if self.have_output:
+            if self.parallel_type == "OpenMP":
+                from gid_output_process import GiDOutputProcess
+                self.output = GiDOutputProcess(self.model_part,
+                                               self.project_parameters["problem_data"]["problem_name"].GetString() ,
+                                               self.project_parameters["output_configuration"])
+            elif self.parallel_type == "MPI":
+                from gid_output_process_mpi import GiDOutputProcessMPI
+                self.output = GiDOutputProcessMPI(self.model_part,
+                                                  self.project_parameters["problem_data"]["problem_name"].GetString() ,
+                                                  self.project_parameters["output_configuration"])
+
+            self.output.ExecuteInitialize()
+
+    def _SetUpRestart(self):
+        """Initialize self.restart_utility as a RestartUtility instance and check if we need to initialize the problem from a restart file."""
+        if self.project_parameters.Has("restart_settings"):
+            restart_settings = self.project_parameters["restart_settings"]
+            self.load_restart = restart_settings["load_restart"].GetBool()
+            self.save_restart = restart_settings["save_restart"].GetBool()
+            restart_settings.RemoveValue("load_restart")
+            restart_settings.RemoveValue("save_restart")
+            restart_settings.AddValue("input_filename", self.project_parameters["problem_data"]["problem_name"])
+            restart_settings.AddValue("echo_level", self.project_parameters["problem_data"]["echo_level"])
+
+            if self.parallel_type == "OpenMP":
+                import restart_utility as ru
+                self.restart_utility = ru.RestartUtility(self.main_model_part,
+                                                         self.project_parameters["restart_settings"])
+            elif self.parallel_type == "MPI":
+                import trilinos_restart_utility as ru
+                self.restart_utility = ru.TrilinosRestartUtility(self.main_model_part,
+                                                                 self.project_parameters["restart_settings"])
+        else:
+            self.load_restart = False
+            self.save_restart = False
+
+    def Solve(self):
```

Comment:
I think it would be good to split this function in several subfunctions.
Otherwise it is not possible to use it for coupled simulations (e.g. FSI)

Reply:
I would propose to name the subfunctions according to the interface of the strategy, that is

* Predict
* InitializeSolutionStep
* SolveSolutionStep
* FinalizeSolutionStep

Reply:
Actually, i believe that we should use a different name rather than Solve, something like SolveTimeLoop

Reply:
+1 for @RiccardoRossi 
Abt the predict I am not sure, I think this could be confusing. E.g. in FSI, would this have to be called once per timestep or once per iteration? => It could be in `InitializeSolutionStep` or `SolveSolutionStep`

Reply:
I agree that some splitting is needed. I also do not see much need for a predict method here. 

However, I think we need to be very clear on what is expected to go in each of the Initialize/Solve/Finalize before settling into this division. Otherwise we might get obscure bugs when different processes/utilities/applications expect to use things that have not yet been properly initialized because (for example) fluids do some thing in finalize and structures do the same thing in initialize.

Reply:
Predict, is definitely needed in some scenarios, although it can be empty in others. 

+1 for agreeing (and casting in stone in the wiki) what should be done where

Reply:
I introduced the subdivisions, I'd propose to take them as provisional for now while we finalize the interface.